tiled/adapters/zarr.py

```diff
@@ -176,13 +174,69 @@ async def write_block(
         -------
 
         """
-        if slice is not ...:
-            raise NotImplementedError
         block_slice, shape = slice_and_shape_from_block_and_chunks(
             block, self.structure().chunks
         )
         self._array[block_slice] = data
 
+    async def patch(
+        self,
+        data: NDArray[Any],
+        slice: Tuple[Union[slice, int], ...],
+        extend: bool = False,
+    ) -> Tuple[Tuple[int, ...], Tuple[Tuple[int, ...], ...]]:
+        """
+        Write data into a slice of the array, maybe extending it.
+
+        If the specified slice does not fit into the array, and extend=True, the
+        array will be resized (expanded, never shrunk) to fit it.
+
+        Parameters
+        ----------
+        data : array-like
+        slice :
+            Where to place the new data
+        extend : bool
+            If slice does not fit wholly within the shape of the existing array,
+            reshape (expand) it to fit if this is True.
+
+        Raises
+        ------
+        ValueError :
+            If slice does not fit wholly with the shape of the existing array
+            and expand is False
+        """
+        current_shape = self._array.shape
+        new_shape = list(current_shape)
+        for i, (s, dim) in enumerate(zip(slice, current_shape)):
+            if isinstance(s, int):
+                new_shape[i] = max(new_shape[i], s)
+            elif isinstance(s, builtins.slice) and isinstance(s.stop, int):
+                new_shape[i] = max(new_shape[i], s.stop)
+            else:
+                raise TypeError(f"Unexpected slice parameter: {slice}")
+        new_shape_tuple = tuple(new_shape)
+        if new_shape_tuple != current_shape:
+            if extend:
+                # Resize the Zarr array to accommodate new data
+                self._array.resize(new_shape_tuple)
+            else:
+                raise Conflicts(
+                    f"Slice {slice} does not fit into array shape {current_shape}. "
+                    f"Use ?extend=true to extend array dimension to fit."
+                )
+        self._array[slice] = data
```

Comment:
If the array was resized, but an exception happens from here on (like someone putting in a bad slicing parameter), the file was resized, but new data not put in. Would we consider a try block that resizes back to the original size?